lib/VM/JSLib/RegExp.cpp

```diff
@@ -460,6 +461,67 @@ static void createGroupsObject(
   JSObject::setNamedSlotValueUnsafe(matchObj.get(), runtime, groupsDesc, shv);
 }
 
+static ExecutionStatus createIndicesArray(
+    Runtime &runtime,
+    Handle<JSArray> matchObj,
+    Handle<JSObject> mappingObj,
+    Handle<JSArray> indices) {
+  if (mappingObj) {
+    // Create groups object and set its prototype to null.
+    auto clazzHandle = runtime.makeHandle(mappingObj->getClass(runtime));
+    auto groupsObjRes = JSObject::create(
+        runtime, Runtime::makeNullHandle<JSObject>(), clazzHandle);
+    auto groupsObj = runtime.makeHandle(groupsObjRes.get());
+
+    // For each capture group, add it to groups object with its corresponding
+    // indices.
+    HiddenClass::forEachProperty(
+        clazzHandle, runtime, [&](SymbolID id, NamedPropertyDescriptor desc) {
+          auto groupIdx =
+              JSObject::getNamedSlotValueUnsafe(*mappingObj, runtime, desc.slot)
+                  .getNumber(runtime);
+
+          JSObject::setNamedSlotValueUnsafe(
+              *groupsObj, runtime, desc.slot, indices->at(runtime, groupIdx));
+        });
+
+    // Add groups object to indices array.
+    if (LLVM_UNLIKELY(
+            JSObject::defineOwnProperty(
+                indices,
+                runtime,
+                Predefined::getSymbolID(Predefined::groups),
+                DefinePropertyFlags::getDefaultNewPropertyFlags(),
+                groupsObj) == ExecutionStatus::EXCEPTION)) {
+      return ExecutionStatus::EXCEPTION;
+    }
+  } else {
+    // If there are no capture groups, then set groups to undefined.
```

Comment:
Here, you can leave an inline comment that copy-pastes the spec. In this case it's coming from this [definition](https://tc39.es/ecma262/#sec-makematchindicesindexpairarray).

```
// 7a. Else, Let groups be undefined.
if (LLVM_UNLIKELY(
            JSObject::defineOwnProperty(
                indices,
                runtime,
                Predefined::getSymbolID(Predefined::groups),
                DefinePropertyFlags::getDefaultNewPropertyFlags(),
                Runtime::getUndefinedValue()) == ExecutionStatus::EXCEPTION)) {
      return ExecutionStatus::EXCEPTION;
    }
```